### ISSUE TITLE
Inconsistent argument conversion when using `None` as default value with Python 3.11 and earlier

### ISSUE THREAD

**Issue by Snooz82** [robotframework/robotframework#4626]
there is something strange happening with Python 3.11 and type hints and auto conversion.
When you have a Type Hint Any  and default Value is None  in 3.11 'none' is always converted to None while in 3.10. it stays a string.
Given that Test:
```robotframework
*** Settings ***
Library    mylib.py

*** Test Cases ***
Test
    Myfunc    none    none

Test 2
    Myfunc    Hello    Hello
```
And That Library
```python
from typing import Any, Optional, Union


def myfunc(value, assertion_expected: Any = None):
    assert value == assertion_expected, (
        "Assertion failed: "
        f"{value} ({type(value)}) != {assertion_expected} ({type(assertion_expected)})"
    )
```
It fails in Python 3.11 and passes in 3.10

`def myfunc(value, assertion_expected: Optional[Any] = None):`
The Optional[Any] fixes it.
Also Libdoc behaves differently and only shows <Any> as Type hint.

**Comment by pekkaklarck** [comment]
The root cause for this issue is pretty interesting. First some details about the conversion process:

1. Type `Any` has no converter.
2. `None` conversion works so that string `NONE`, case-insensitively, is converted to `None` object.
3. If conversion done based on a type hint fails, a type hint has no converter, or there's no type hint in the first place, Robot tries conversion based on possible default value.
4. Robot has special `CombinedConverter` for unions. It tries converters for each nested type separately. If a converter can handle the value, the converted value is returned. If a type doesn't have a converter, the value is returned as-is. If all types have a converter and they all fail, the whole `CombinedConverter` fails.
5. Prior to Python 3.11, [typing.get_type_hints](https://docs.python.org/3/library/typing.html#typing.get_type_hints) that we use for getting type information returned same info with `arg: X = None` and `arg: Optional[X] = None`. The latter is just syntactic sugar for `arg: Union[X, None] = None`.
6. With Python 3.11, `arg: X = None` has no special handling and `typing.get_type_hints` returns just `X`.

Taking the above into account, this is what happens with Python 3.10 when you use `arg: Any = None`:
1. The syntax is handled the same way as `arg: Union[Any, None] = None`.
2. `CombinedConverter` is used.
3. `Any´ has no converter so `ConbinedConverter` returns the used value as-is.
4. Conversion succeeded so the default value is not used for anything.

And this is what happens with Python 3.11:
1. The syntax isn't anymore considered same as using `Union`.
2. Conversions is done using `Any` directly.
3. `Any` has no converter so Robot look for default value.
4. `None` conversion converts string `none` to `None` object.

**Comment by pekkaklarck** [comment]
This issue makes it clear that conversion based on default values works differently when the type is a union compared to when there's just a single type:
- If you use just `arg: T = default` and `T` has no converter, conversion is attempted based on the type of the default value.
- If you use `arg: Union[T, S] = default` and `T` has no converter, the value is always used without conversion.

This inconsistency could be fixed so that `CombinedConverter` ignores types that don't have a converter. That would make the behavior with `arg: Any = None` consistent with different Python versions so that also older versions would work like Python 3.11. That would be a backwards incompatible change and thus not something we'd like to do in RF 6.1. I'm also not sure would that change be a good idea in general.

Another way to handle this inconsistency would be not looking at default values at all if an argument has a type hint, regardless does that type have a converter or not. Also this would make the behavior with  `arg: Any = None` consistent, but now Python 3.11 would work the same way as older versions. I think that's a better solution than the earlier, but also it is backwards incompatible and thus not desirable in RF 6.1.

I believe the best way to handle this particular issue now is handling `None` as default value specially with Python 3.11+ so that it works the same way as with earlier Python versions. We could also consider not looking at default values at all if there's a type hint, but that requires a separate issue and a deprecation period.

**Comment by pekkaklarck** [comment]
Good thing with this issue is that changing `arg: Any = None` to `arg: Optional[Any] = None`, `arg: Union[Any, None] = None` or `arg: Any|None = None` (my favorite but requires Python 3.10) fixes the issue and works the same way with all Python versions. I thus don't consider this issue high priority. Needs to be fixed in RF 6.1. anyway.

**Comment by Snooz82** [comment]
Should we assign milestone 6.1?

**Comment by pekkaklarck** [comment]
Yeah, inconsistency in behavior with different Python versions can, and should, be fixed in 6.1. I already looked at that and it doesn't look too hard.

Inconsistency in conversion based on default values needs a separate issue and due to backwards incompatibility reasons needs to wait until RF 7. It also require deprecation latest in RF 6.2, preferably already in RF 6.1.

**Comment by pekkaklarck** [comment]
This ought to be fixed now. Needed to also take it into account that with `arg: T = None` we actually do `None` conversion, also with older Python versions, if `T` is a known type but conversion fails. For example, `arg: list = None` works that way and breaking it would be backwards incompatible.

As I wrote earlier, it probably would be best if we didn't look at default values at all if an argument has type information. That would obviously mean that behavior with the aforementioned `arg: list = None` would change, and thus the change could only be done in a major version and with a deprecation period. Users basically needed to change that syntax to `arg: Union[list, None] = None` or, if they use Python 3.10+, to `arg: list|None = None`. I personally don't like the former syntax too much and it also requires adding `from typing import Union`. Perhaps we should go with the current functionality now and think about this again later.
